### ISSUE TITLE
Sage x86 SSE code seems broken with DJGPP

### ISSUE THREAD

**Issue by maraakate** [dborca/sage#1]
I know it's stated no more work will be done from dborca himself, but this is a serious issue that someone using this project must know about.

Dynamic DXE builds with Sage usually fail to initialize and the program segfaults (but the screen is black) on SSE processors (i.e. Pentium 3 and later)

Some stuff I've used it with I can get it start by passing a specific string length range of junk paramters at startup. i.e. q2.exe -thisisalongstring

For reference, the actual offending code is in sage\drivers\glide\x86\sse_vbtmp.asm. if sse_emitvertices friends are bypassed in sage\drivers\glide\drv_vb.c then the problem doesnt occur but then performance is basically the same as if SSE was disabled.

proc TAG(sse_emitvertices)

See our issue at Q2DOS for more information:
https://bitbucket.org/neozeed/q2dos/issues/12/sage-x86-sse-code-seems-broken

**Comment by sezero** [comment]
For dxe builds of sage, we are using the fork by sezero, here: https://github.com/sezero/sage

**Comment by dborca** [comment]
I have read the story from bitbucket. So: 1) it seems to be like a 50% occurrence, 2) static builds are unaffected and 3) passing a specific string length range of junk parameters at start-up seems to "fix" it.

It looks like a memory corruption to me... in any case the causal chain may be longer than it seems. The "parameter fix" is not a fix, it just hides the problem by changing the order in which memory blocks are allocated, etc.

Graphics stuff is hard to debug in DOS. On top of that, you probably won't get a meaningful back-trace in SSE code, because it seems I sacrificed the frame-pointer register (ebp).

Your best bet is to try to do some printf-debugging. Just pepper the code with printfs and observe the output. It sounds silly, but it usually works. I, myself, cannot debug it because I don't have any rig available. At most, I _could_ review the code, but I'm really short on spare time.

Alternatively, you could try to run the Linux version through valgrind (that is, if valgrind supports SSE now).  Back in the days I didn't have valgrind, so I used some other memory checker: MSS. Try to spot the bug there; any bug.

Also, make sure the structs are tightly packed. Maybe the defines in drv.inc are not reflecting the actual layout of the respective structures? These can vary from compiler to compiler and version to version.

Also, if you're building with FX_PACKEDCOLOR, I think SSE code is mixed with 3DNow! code, so pay attention to emms/femms for your target processor.

Last, but not least, I know DXE modules were pretty limited, so maybe they are the source of your problem (see #2). I can't remember who designed them (maybe Charles W. Sandmann?) but I do remember DXE limitations were documented somewhere. If that is the case, you'll need to work around them. As a side-note, that is exactly the reason I wrote DXE3 -- together with an self-relocatable DJELF toolchain -- but that didn't get any traction (although cws liked the idea).

**Comment by maraakate** [comment]
Thanks for the reply Daniel,

The static fix is not a fix... it's just a workaround I found for myself 
personally.  We're not advertising that to potential users. The 
occurence is more than 50% I'd say more like 75%.  Sometimes you can 
keep spamming q2.exe -> enter -> up (doskey will retype q2.exe) -> enter 
... etc.

I figured it was an actual bug in the SSE ASM code; and I have no clue 
how to spread a printf to this.  Though I am slightly confused by the 
removal of that makes it work OK again because I assume emitvertices 
isn't going to be triggered right away as soon as the game is started to 
the console backdrop.

Static builds have a different issue altogether.  I don't remember if 
SSE has to enabled for it but you seem to get like 2 or 3 mode switches 
per session then it dumps to DOS again.  I haven't retested this issue 
since we removed the atexit for MSDOS code path so that may have been 
part of it.

Another general goofiness I found was that the init needs to happen 
twice to get the proper performance it's capable of.  Sezero suspects 
that may be related to him not doing the right order, but it appeared OK 
to me.  The difference is drastic, i.e. like a 20fps on average on a p3 
900 machine in running timedemos in Quake 2. It's not a bug affected by 
SLI, which I initially suspected because earlier Mesa 5 builds had 
issues with not setting SLI properly until it was reinitialized again.

I'm not using 3Dnow.  I'm pretty much the guinea pig for this particular 
project and I don't own anything AMD.  I did try rebuilding with 
FX_PACKEDCOLOR at one point and it was not that.

DXE appears to be very capable these days (I was quite surprised at that 
myself).  We are using the DXE's just like Quake 2 does with the DLLs 
and ref_soft is OK and Mesa with ref_gl is OK.

Thanks for you input,
Frank

**Comment by ghost** [comment]
Is it possible to show the page fault error output? Does the issue exist for a Windows build? Was the build tested on a Pentium4 along with the corresponding -march flag?

"Also, if you're building with FX_PACKEDCOLOR, I think SSE code is mixed with 3DNow! code, so pay attention to emms/femms for your target processor."

Verified sse_vbtmp.asm has the instruction:
%if FX_PACKEDCOLOR
    emms

However, if FX_PACKEDCOLOR is turned off by default, then the instruction is not called.

**Comment by maraakate** [comment]
I'm unsure if the issue is DOS isolated.  I don't have a Pentium 4 machine to test it on.  Well, technically I do somewhere but it's in storage it may not even work anymore.

The trace from symify isn't very useful because what happens is WATT32 has a signal handler installed and WATT32 is catching it before it locks the machine up.  If you use WATT32s environment variable to disable the signal handler (forgot what it is specifically, you'll have to read the docs or look at the source) then the machine hard locks instead.

I don't have a proper mingw or any Linux cross compile tools.  I'm compiling with DJGPP 2.05 and GCC3.46 on Windows XP 32-bit.  If someone can compile a new version based on sezeros source for win32 I will test it.

**Comment by ghost** [comment]
Is the keyboard input to the console also frozen, or just that the screen is black? If the latter, then it may be possible to type the symify command line even though the console is not seen.

**Comment by maraakate** [comment]
Also, for shits and giggles I'll retest compiling with FX_PACKEDCOLOR removed.  The -march stuff is irrelevant for Quake 1 and 2 under DOS.  I tried compiling the games themselves with -march=pentium3 and omitting a -march flag entirely and it makes no difference with this particular bug.  I've never tried compiling sage itself with -march=pentium3 as we're trying to aim for this working across multiple configurations right now

**Comment by maraakate** [comment]
It's not frozen normally, when the game goes to black screen it's actually just that the context isn't freed and reset properly so yes you can get a symify dump.  I have done this before by typing the batch file then just starting another DOS program that resets the video and stuff works again.  As I said, the only time that doesnt work is if signal handler is disabled from WATT32 then it's a legitimate hardlock with a black screen and no background typing will fix that.

**Comment by ghost** [comment]
I think some of the older P3 models support single precision SSE only, but it does seem like a memory issue as noted above instead of the direct cause being the SSE code.

**Comment by maraakate** [comment]
Memory issue in Sage, Quake 2 or Glide3x, or even something with DJGPP and DXE?  I'm not sure how one narrows it down.  Otherwise, when SSE code is bypassed everything works properly, you can switch mods and change resolutions all day long.

**Comment by ghost** [comment]
Also, is it possible to remove the watt32 code and test again?

**Comment by ghost** [comment]
I think the only easy way to test is the symify dump.

**Comment by maraakate** [comment]
The symify dump may be useless because afaik there is no real debugging from DXEs, but I can retry building the DXEs with -g and optimizations off and see what happens.

**Comment by ghost** [comment]
It would rule out that the core q2 code is not the likely cause, or at least provide a possible clue.

**Comment by maraakate** [comment]
Well the odd issue I was having with vid_restarts in static builds was Hunk_Alloc failures when loading a new map.  I'll revisit that again too.

**Comment by maraakate** [comment]
And the Hunk_Alloc static build thing was isolated to sage static builds.

**Comment by ghost** [comment]
Did you try reverting the hunk_alloc/realloc code?

**Comment by maraakate** [comment]
Revert what aspect of it?  So much stuff has changed I really don't even remember what all I have done or sezero.

**Comment by maraakate** [comment]
But, I figured if the Hunk_Alloc code had legitimate problems it would have appeared a long time ago because I've tested the game to hell and back playing xatrix, rogue, zaero, weapons mod online (that huge nutty Wod X server), single player campaign, hell even large single player map series like 1964 over Mesa and ref_soft and all that stuff works fine and I'd always switch modes every now and again because 640x480 helps a lot in large deathmatches or huge maps that really just cant handle anything higher on my machine.

**Comment by ghost** [comment]
What typically causes a hunk_alloc error? Does any kind of -mem parameter, is -heapsize present?, or another parameter affect that error?

**Comment by maraakate** [comment]
I don't think Quake 2 has a concept of mem or heapsize.  There is no 
actual locking of the entire RAM down like Quake 1 with it's custom 
memory stuff.

The problem is irregardless of using CWSDPR0.EXE or CWSDPMI.EXE with 
Virtual Memory.  And this is happening on relatively large RAM machines 
(like 512MB to 1GB, way more than Q2 needs).

The crash was coming from ref_gl where it uses Hunk_Alloc which may be 
related to something in ref_gl from Knightmare.  There is a current 
issue with ref_gl during the Mod_AliasLoadModel (or similar titled 
function name) and chaos mod with a certain model with a hard lock and 
the hard lock happens on Windows as well (well it hard locks the 
application not the entire OS)... so there may be something funky in 
Knightmare's revised ref_gl driver.

Frank

**Comment by ghost** [comment]
Those bugs seem like tractable problems. Those two issues may be worth solving first and then revisit a possible memory issue leading to the SSE problem. Did you try increasing/decreasing the memory available for the stack by cwsdpmi?

**Comment by maraakate** [comment]
However the SSE crash with Sage persists in Quake 1 as well and I always start that game with -mem 64 so I have way more than enough space to play with.

**Comment by ghost** [comment]
That's interesting. However, if a Windows build works, then the DOS memory model is suspect.

**Comment by maraakate** [comment]
Yeah, I tried a ridiculous amount like 8MB and that made no difference.  I also tried it in barebones DOS bootdisk to be sure it wasn't anything funny like a TSR.

**Comment by maraakate** [comment]
Also tried setting statcks=0,0 and other nonsense.  None of changes the issue at hand.

**Comment by ghost** [comment]
Did you try different stack sizes with stubedit only? (assuming no code altering the value in q2)

**Comment by maraakate** [comment]
However, just because Windows allows it doesn't mean linux might.  I may just need to bite the bullet and setup a linux box to really debug this.  I set the stacks directly in sys_dos.c.

A case in point with windows allowing awful code is dosdoom had some overhauled crazy saving system and it only works in a windows 98 dos box.  If you play it on a real dos computer restoring saves has null pointer crashes, which means they probably only did their development on Windows 98 dos boxes and started the game up in DOS to make sure it actually worked and assumed all was OK.  Was never able to track that bug down.

**Comment by maraakate** [comment]
And for kicks I tried compiling everything with GCC4.84 to see if it was some compiler issue (we recently had another issue with SLI corruption from GCC3.46 when used with -O2 for the Voodoo 5 driver) but no dice.

**Comment by ghost** [comment]
Heh.

Can you try changing the stacks value in sys_dos.c, both up and down by 512k?

**Comment by maraakate** [comment]
Can having too much stack be a problem?

**Comment by ghost** [comment]
If the error is not deterministic, then it is worth ruling out that the memory is not an issue at some level, even if in the stack memory area.

**Comment by maraakate** [comment]
Anyways, the value we're currently using is 1048576, what should I try setting it to?  Something like 524288 and 1572864?

**Comment by ghost** [comment]
Exactly, see if it changes the result in any way.

**Comment by maraakate** [comment]
Heh, 524288  is a hard lock :)

**Comment by ghost** [comment]
:)

**Comment by maraakate** [comment]
I'm not sure what you want me to test specifically, but this is a static sage build I'm trying these values on.  1.5mb is still crashing on mode changes.  It only happens after a map has been loaded during the life time of the program.

**Comment by ghost** [comment]
It is good to make sure. Also, on the SSE related issue, it may be worthwhile to inline a portion of the code into the C path for that function. Inline pieces until it breaks or the performance is up to snuff.

**Comment by maraakate** [comment]
The ASM code itself is some unrolled loop stuff I believe and I really have no clue what I'm doing with ASM but we may have to end up going that route.  I think it used some clever macro stuff too.

**Comment by ghost** [comment]
Good point. Then the symify provides the easier path to the error (in the hope that it can find the true cause in the code). It may also help to test a Windows binary at some point.

**Comment by ghost** [comment]
May also be worthwhile to remove knightmare's additions from the code and testing, too.

**Comment by ghost** [comment]
Do any of the archived sage binaries reveal which gcc compiler version was used? Perhaps in any linux build?

**Comment by maraakate** [comment]
I don't know, daniel would know the answer to that.  Knightmare did a lot of changes, so the 3.20 ref_gl will have to ported to q2dos again (there was a separate branch I think I kept around for it, but it may be deleted by now).

**Comment by ghost** [comment]
In the worst scenario, some of the SSE code can be inlined into the C path for the function with the issue. However, the above may help. It is good to know that qdos has the same issue with sage, so it does suggest something in the gl code or sage itself (or the dxe3 system), but any symify luck would help choose the more likely source.

**Comment by ghost** [comment]
Do the demos fail with sage, too?

**Comment by maraakate** [comment]
What do you mean the demos?  But in any case the traceback for the Hunk_Alloc issue relates to Mod_ForName and then Hunk_Alloc and finally its malloc.  It crashes otu with combinehi: memory fouled.

**Comment by ghost** [comment]
The hunk_alloc issue should be tractable, at least its not a semi-random error. Also, the demos from mesa should build with sage (with some modification), much like you did with quake1/2. It is another way to test the SSE code.

**Comment by maraakate** [comment]
Oh, I've never tried any of those demos.

**Comment by ghost** [comment]
Is the mod_forname problem also seen in the win32 build?

**Comment by maraakate** [comment]
I have a feeling the Hunk_Alloc thing is part of some deeper issue possibly with how Sage may be setting up contexts or other memory stuff that is screwing things up.

**Comment by maraakate** [comment]
No, that's using a different memory manager scheme entrusted to windows.  DJ 2.05 had redone malloc and friends so its quite possible there is a bug there or in sage fouling things up.

**Comment by ghost** [comment]
I agree. The program probably spends a good deal of time in that sage function, too.

**Comment by maraakate** [comment]
I also have no sage dll for windows based on this code to test.

**Comment by ghost** [comment]
Hopefully the mesa demos will substitute for a sage dll, although Daniel may have released a binary for win32.

**Comment by maraakate** [comment]
There was binaries for win32 out there on his geocities page but it was for voodoo 2 only and I couldn't get it to work on my Voodoo 2.  I think sezero built me a win32 build to test on my voodoo 5 recently, I can revisit that again.  But in any case, the Hunk_Alloc -> malloc issue doesn't care about SSE being enabled or not so it's not the SSE ASM code.

**Comment by maraakate** [comment]
That's right I tried this binary sezero sent me previously and win32 will crash if I try to set it in windowed mode, but if I do vid_fullscreen 1 it drops back to ref_soft because it says choosepixelformat failed for 640x480

**Comment by ghost** [comment]
Very good. Hopefully the mesa demos show an error with the SSE code active and while not statically linked to the library, so the error can be isolated further.

**Comment by ghost** [comment]
Are there a couple of functions in sage to disable where using win32?

**Comment by maraakate** [comment]
Well, afaik sage when used with win32 is supposed to be emulating how windows setups contexts for the opengl window.  In q2dos we are calling the sage_init and sage_bind directly for creating their contexts.  If you're able to rebuild sage from sezeros branch as a dll and send it to me I can retest.

**Comment by ghoul8765** [comment]
# Ignore alphabuffer requests in wglChoosePixelFormat
# Not recommended, unless you want to force certain games to boot
# eg: Doom3, Kotor1/2
#wgl.ignore.alphabuffer		= y

# Ignore stencilbuffer requests in wglChoosePixelFormat
# Not recommended, unless you want to force certain games to boot
# eg: Doom3, Kotor1/2, NeverwinterNights
#wgl.ignore.stencilbuffer	= y

**Comment by ghost** [comment]
Ignore stereo requests in wglChoosePixelFormat.
wgl.ignore.stereo       = y

 Ignore alphabuffer requests in wglChoosePixelFormat
wgl.ignore.alphabuffer      = y

 Ignore stencilbuffer requests in wglChoosePixelFormat
wgl.ignore.stencilbuffer    = y

**Comment by maraakate** [comment]
Yeah, I think I tried those before and it didn't help but I will retry again just so we can have that ruled out.

**Comment by ghost** [comment]
I'll work on a setup to build the dll.

**Comment by maraakate** [comment]
Still doesn't work.  Dunno what to say there.

**Comment by ghost** [comment]
Are the Pentium3 and fpmath=both optimizations approaching the speed of the SSE code path? I understand this method, even if it helped, would require multiple binaries for the user.

**Comment by maraakate** [comment]
I dont force fpmath, i just use march=pentium3 on Sage when I tested previously and GCC generated code similar in speed to the SSE ASM daniel wrote.  There was about a 5 fps penalty hit, but with SSE code on I usually gained an extra 20fps on most scenes and timedemos in Q1 and Q2.  So that would still be 15fps gained from not using anything and still would make it faster than Mesa.

**Comment by ghost** [comment]
So either way you have a gl library that is faster than Mesa. :)

**Comment by maraakate** [comment]
With no SSE optimizations Mesa will beat Sage as Mesa has it's own SSE code it uses as well.  I've never tried Mesa with the SSE code bypassed.

**Comment by maraakate** [comment]
But requiring a separate gl.dxe that's compiled as SSE for the user is a bad thing and is just dancing around the problem.  And in all fairness I'm the only person out there developing and testing this on real hardware.  The vogons clowns were trying all this stuff on dos box with glide which isn't much help as dosbox will allow bad things to happen and mask the errors (like when we had nearptr issues very early on with SVGA).

**Comment by ghost** [comment]
That sounds impressive.

**Comment by maraakate** [comment]
I recall I lived extra dangerously ;) and compiled a march=pentium3 with SSE ASM optimizations of Sage and that had no real true measurable difference between choosing one or the other.

**Comment by ghost** [comment]
Best to test on real hardware. Perhaps the CTTY command can allow debug output via the serial port?

**Comment by maraakate** [comment]
I do have null modem cables and extra CRTs and such (presently I have 3 separate machines stacked next to each other with a KVM to test between voodoo 2, voodoo 3, and voodoo 5 for this project and I have a voodoo 1 and rush arriving soon from falconfly for more testing as well as PCI carsd from him so I can test on modern machines).  So assuming the serial output debugging is maintained in DJGPP it should be possible.  I don't use much GDB I'd need a bit of hand holding on it.  Visual Studio's debugger makes it too easy to find fuck ups compared to GDB.

**Comment by maraakate** [comment]
The Voodoo 2 computer is a P1 200 mmx and everything works properly on that computer.  BUt, I have not tried a static build of sage on that particular machine to see if the Hunk_Alloc issue persists there.  The V3 and V5 machines are Pentium 3 machines.

**Comment by ghost** [comment]
It provides another avenue to debug the issue. :)

**Comment by maraakate** [comment]
I take it you don't live near me, I'm in Lancaster, PA, USA, because having someone around to assist with it would make my life easier instead of copy and paste or having to type stuff out and waiting on what I should input next.

You never know, sometimes it really is a small world so it's worth asking. haha.

**Comment by ghost** [comment]
:) I think the CTTY would help if the printf commands were becoming burdensome.

I'll work on a build environment for the sage dll.

**Comment by maraakate** [comment]
I guess it's worth noting that since the Voodoo 3 with a P3 exhibits the same problems we can rule out that it's an SLI issue with Voodoo 5 5500 or any 32-bit colour stuff.

**Comment by ghost** [comment]
Very good. Also, if you had a menu option, then would it be possible to include an SSE and a non-SSE optimized gl.dxe, where the user selects the menu item instead of replacing files?

**Comment by maraakate** [comment]
That's still just dancing around whatever fundamental problem is lurking.  Right now the compromise is that SSE ASM is disabled by default in our sage.ini and has to be explicitly turned back on.  As stated, you can pass junk parameters of varying length and get the program to work fine that way.

**Comment by maraakate** [comment]
But yes, there is plans at some point to implement gl_driver from win32 to allow switching between Mesa and Sage on the fly.

**Comment by ghost** [comment]
I wonder why the junk parameters avoids the issue. I noted Daniel said it will alter the memory setup somewhat (hopefully paraphrased correctly).

**Comment by maraakate** [comment]
Possibly related to our Hunk_Alloc corruption resulting from static builds with mallocs down the line.

**Comment by ghost** [comment]
I think you have the issue reasonably isolated given the tests done. I hope a mesa demo will mirror these issues for testing.

**Comment by ghost** [comment]
I forget if the hunk_alloc corruption occurs in qdos or just q2dos.

**Comment by maraakate** [comment]
Well, at some point here I will use quake 2 3.20 ref_gl and see if the issue with static builds go away there.

**Comment by ghost** [comment]
I hope that fixes it. :)

**Comment by maraakate** [comment]
Hunk_Alloc corruption is currently just a q2dos issue as my QDOS build does not yet have dynamic video modes.

**Comment by maraakate** [comment]
But QDOS still has the SSE at startup issue just like Q2DOS.

**Comment by ghost** [comment]
I think the hunk_alloc issue can be more easily isolated. I may be able to test that at some point, too.

**Comment by maraakate** [comment]
For all we know it may magically fix our memory corruption issues that daniel thinks are going on with Sage and DOS.

**Comment by ghost** [comment]
I hope so, too.

Also, for a temporary workaround, it may be possible to create a copy of the current C function that corresponds to the problematic SSE function. That copy can be surrounded by pragma hints for Pentium3, fpmath=both, and/or fast-math. That copy would be accessed where the sage.ini indicates SSE=yes while the original function is accessed where SSE=no. That would retain a single gl.dxe while allowing both paths. This assumes that the pragma hints and the working SSE code continues to approximate the full SSE path in performance.

**Comment by maraakate** [comment]
for the record, fast-math makes no difference.  Even with Quake on ancient machines.

**Comment by maraakate** [comment]
That's basically what I was thinking of doing at some point, except 
taking the *.s files.  But your way makes more sense, definitely less 
painful and GCC would still be trusted to do it.  Except fpmath.  Sezero 
has told me that can be a bad thing to do.  I used to do fpmath=sse or 
whatever it's supposed to be for such things and it made no speed 
differences in quake 2 as opposed to the possibly safer march=pentium3 
flags.  How correct all this is, I don't know. But sezero hasn't steered 
me wrong so I trust his judgement on minimizing potential subtle bugs.

I can try recompiling later with the sse_emitvertices codepath disabled 
from using the asm file, but still use march=pentium3 and compare 
performance.  But we're jumping ahead here.  First we need to find out 
why Sage is not working for me on Windows 98SE (have not tested any of 
this on XP).  Mesa DLLs work as they should on Win98SE so I assume Sage 
should do the same.  Then see if some of these issues persist there and 
if they do it would probably make life easier to debug from Windows.

Frank

**Comment by ghost** [comment]
I agree entirely -- just suggesting a variation on the theme of optimizations.

I'll continue the mingw work to build a dll which would work on 98. This is an important test.

**Comment by maraakate** [comment]
Your method is a lot better than the pain I would have submitted myself 
to.  I've only been programming as a hobby for about 4-5 years and this 
is the most complex project I've worked on besides Daikatana 1.3 so I 
don't always know the most efficient way to attack a problem :P

Frank

**Comment by ghost** [comment]
That method was a reflection of yours to use pragma hints. I only hope that it helps. Also, your projects are quite extensive in scope and code. It is complex and difficult to work across the different subsystems and implement video, audio and all the supporting code.

While working on mingw, I did take a quick look at fxtris.o and the cross-compiler issue. I think the non-working copy has rearranged/additional "nop" and nop-like instructions in the stencil code and possibly later in the code. I wonder if the stencil code is used in your projects, but the cross-compiler seems problematic. On the other hand, the older binutils are presumably producing good code, where compiling on the target, but in this case also produces different code than the referenced object file.

**Comment by maraakate** [comment]
> That method was a reflection of yours to use pragma hints. I only hope 
> that it helps. Also, your projects are quite extensive in scope and 
> code. It is complex and difficult to work across the different 
> subsystems and implement video, audio and all the supporting code.
> 
> Yeah, especially when we're relying on other projects so we don't have 
> to reinvent the wheel (i.e. cURL, mpxplayer for PCI audio, glide3x, 
> sage, mesa and WATT32).  And then on top of that we're also having the 
> renderer and game logic as DXEs just like Q2 originally works so there 
> doesn't need to be a separate build of every mod, etc.
> 
> While working on mingw, I did take a quick look at fxtris.o and the 
> cross-compiler issue. I think the non-working copy has 
> rearranged/additional "nop" and nop-like instructions in the stencil 
> code and possibly later in the code. I wonder if the stencil code is 
> used in your projects, but the cross-compiler seems problematic. On 
> the other hand, the older binutils are presumably producing working 
> code where compiling on the target but also produces different code.
> 
> We're not using any stencil buffers, though I think Knightmare's code 
> actually supports stencil buffers for shadows but a Voodoo card isn't 
> going to be able to handle that anyways so it's purposeless to declare a 
> stencil buffer during the context initialization.

Frank

**Comment by ghost** [comment]
That's impressive!

Also, you may be able to circumvent any cross-compiler issues by defining away some unused functions.

The sage dll is built but I didn't define a SSE path yet, at least explicitly.

**Comment by maraakate** [comment]
Well, the point of asking daniel to release the sage source was so we could have a stripped down gl driver for DOS.  Sage is more like a mini gl driver where it has the functions it needs to play games and cuts down on file size.

I can test the sage.dll, you can either attach it here or upload it somewhere.  Did you build with X86=1?  This is important as it will enable the detection of processor features like 3DNow! and SSE and use their ASM friends.

**Comment by ghost** [comment]
I'll rebuild it with x86=1. I did see that in the makefile just now.

**Comment by maraakate** [comment]
make sure its X86=1 not x86=1, it's case sensitive.

**Comment by ghost** [comment]
I verified its correct via the command line output (X86=1).

**Comment by maraakate** [comment]
Ok, cool.  I've made those mistakes in the past testing Mesa versions when I was tired and typing quickly ;P

**Comment by ghost** [comment]
:) Fortunately the binary built easily. I didn't test it, but I'll upload it and post a link.

**Comment by ghost** [comment]
I have an account at vogons that I haven't used yet. Are you active there where I can PM the file?

**Comment by maraakate** [comment]
Yes, http://www.vogons.org/memberlist.php?mode=viewprofile&u=23522

**Comment by ghost** [comment]
Thanks. It's uploaded. :)

**Comment by maraakate** [comment]
Don't know why, but ChoosePixelFormat still fails.  I'll have to retry 
it later on my other computer that has Voodoo 3 and Windows XP.

You may even be able to test this with a glide wrapper under Windows 
since sage relies on glide3x.dll to function.

Frank

**Comment by ghost** [comment]
That's cool about the wrapper. I'll try nglide and glide3x.dll on windows then. I hope the win32 sage helps.

**Comment by ghost** [comment]
It shows the same issue about choosepixelformat. I think I've seen that occur in Q3....

**Comment by maraakate** [comment]
I might just recompile the windows gl code that creates the context to 
just use sage_init and sage_bind like dos does and see if that works.

Frank

**Comment by ghost** [comment]
I think the glimp.c may just require an edit.

**Comment by maraakate** [comment]
Yeah, I couldn't remember the filename offhand, but yes that's it. If 
you're up to it you can compare the dos_sage.c file on how the context 
gets created and see if windows will allow it.

I compile it with VS6.  I tried VS2005 but theres OpenGL rendering 
problems with 2005 for some reason (under modern computers, never tested 
it on voodoo in windows)

Frank

**Comment by ghost** [comment]
Maybe some flag like this "PFD_GENERIC_FORMAT" or "PFD_GENERIC_ACCELERATED" for ChoosePFD?

**Comment by maraakate** [comment]
I have no clue.  I never looked at the Windows Sage code.

Frank

**Comment by ghost** [comment]
In glw_imp.c:
gl_allow_software

I think that value can be changed or at least the error bypassed in the q2dos/win32 file.

**Comment by maraakate** [comment]
That just makes it use the ungodly slow microsoft software opengl 
implementation.  But to humour you, I tried it anyway and it still 
doesn't work.  I'll try recompiling ref_gl later today with the sage 
contexts instead and see if that works.

Frank

**Comment by ghost** [comment]
It's here I think:
        if ( ( pixelformat = ChoosePixelFormat( glw_state.hDC, &pfd)) == 0 )
        {
            ri.Con_Printf (PRINT_ALL, "GLimp_Init() - ChoosePixelFormat failed\n");
            return false;

**Comment by maraakate** [comment]
right, so now you have to trace that back and see what the deal is in sage.

frank

**Comment by ghost** [comment]
In q3 the error can be bypassed by r_allowSoftwareGL 1 and it alters few lines in the init code. I'm hoping q2 is similar.

**Comment by ghost** [comment]
```
    if ( ( pixelformat = ChoosePixelFormat( glw_state.hDC, &pfd)) == 0 )
    {
        ri.Con_Printf (PRINT_ALL, "GLimp_Init() - ChoosePixelFormat failed\n");
        return false;
```

Could try to remove "return false;"

**Comment by maraakate** [comment]
The better thing to do would be to look at wglChoosePixelFormat in Sage 
itself and see where it's failing.  Should be able to use printfs, start 
quake2 from cmd.exe with quake2.exe >out.txt and then look at out.txt to 
get your printfs

Frank

**Comment by ghost** [comment]
Good idea. I've been trying to run the vanilla mesa demos against opengl32.dll, but it requires glut, too.

**Comment by maraakate** [comment]
I would just put a printf before every return in those switch 
structures, continues, etc. in there and see where it's blowing up. It 
looks like its passing proper flags to it but maybe it doesnt like 
knightmares changes of defaulting to 32-bit color depth instead of 
16-bit and using a 24-bit z-buffer with 8-bit stencil buffer instead of 
32-bit and 0 respectively.

Frank

**Comment by ghost** [comment]
I wish I had a VS compiler installed. I can install one in a day or two and then test your suggested code changes.

**Comment by maraakate** [comment]
Keep in mind, I have no idea how this actually works in windows land, 
but theres a table for PIXELFORMATDESCRIPTOR pfd[] , line 30 in wgl.c

Maybe adding the new one knightmare has at the end would help it work again:

```
 {
     sizeof(PIXELFORMATDESCRIPTOR),    // size of this pfd
     1,                                // version number
     PFD_DRAW_TO_WINDOW |            // support window
     PFD_SUPPORT_OPENGL |            // support OpenGL
     PFD_DOUBLEBUFFER,                // double buffered
     PFD_TYPE_RGBA,                    // RGBA type
     // Knightmare- was 24-bit, changed for stencil buffer
     32,                                // 32-bit color depth
     0, 0, 0, 0, 0, 0,                // color bits ignored
     0,                                // no alpha buffer
     0,                                // shift bit ignored
     0,                                // no accumulation buffer
     0, 0, 0, 0,                     // accum bits ignored
     // Knightmare- stencil buffer, was 32-bit z-buffer, no stencil 
```

buffer
         24,                                // 24-bit z-buffer
         8,                                // 8-bit stencil buffer
         0,                                // no auxiliary buffer
         PFD_MAIN_PLANE,                    // main layer
         0,                                // reserved
         0, 0, 0                            // layer masks ignored
     }

**Comment by ghost** [comment]
Thanks!

Also, the sage dll should work with Q3 where  r_allowSoftwareGL 1. At least it would show that the dlls are working, I think. Or possibly Glquake.

**Comment by maraakate** [comment]
I just recompiled ref_gl with a quick check of using one of the tables 
from wgl.c and still no dice, so printfs will have to be sprinkled in 
there for me to find out where it's failing specifically.

Frank

**Comment by ghost** [comment]
You're right, I tried glquake and it specifically asked for glColor3ubv in the dll.

**Comment by maraakate** [comment]
It must be stubbed or something because glquake works under DOS with 
Sage (and quite well for the most part, heavy particle tests like 
bigass1.dem lose to Mesa in a sad way)

Frank

**Comment by ghost** [comment]
That's interesting. I found the function in the dll, but the exe can't see it.

**Comment by ghost** [comment]
Replaced with a mesa653 dll and glquake started right up.

**Comment by ghost** [comment]
http://www.geocities.ws/dborca/opengl/sage.html

**Comment by ghost** [comment]
That sage dll from Daniel's web site works with Q2 but not glquake. At least it shows that the dll should work without pixel error and without modifying game code.

**Comment by maraakate** [comment]
Right Mesa DLLs work fine.  It's something with Sage and Quake 2 and 
wglChoosePixelFormat that needs narrowed down before we can see if these 
issues persist in windows land also.  So if you can recompile me a DLL 
with printfs all splattered in that function in sage then I might be 
able to help narrow down why it's failing that call.

Heh, the 0.9.2 crashes immediately for me.  Completely exits the 
program.  Hopefully from the SSE code issue.

Frank

**Comment by ghost** [comment]
:) I guess I must have the SSE turned off.

**Comment by maraakate** [comment]
I tried using the sage.ini to control that with your build and it doesnt 
matter either way there.

I'll have to try it with a Pentium 2 later and see if it still crashes, 
but in the meantime if you can get me a new dll with the printfs id 
appreciate it.

Frank

**Comment by ghost** [comment]
I'll continue to work on this. It may take a while, but at least there is some progress.

**Comment by sezero** [comment]
Are you using Daniel's source as is, or do you have a fork?

**Comment by ghost** [comment]
I used your fork and verified some commits surrounded by dos definition lines.

I'm away from the compiler, but my next guess is that Daniel's sage dll has strings to identify itself as a full opengl icd. And if so, then the source may not have this so Q1/2 is ignoring the library (I think in glw_imp.c). (Another scenario is that the dll is a malformed library, but I haven't yet tried to link against it with a demo nor tried Q3 with the parameter to use the software rasterizer.)

**Comment by dborca** [comment]
Here's how I'd do the printf debugging from inside the ASM sources: http://pastie.org/10450076

%include "xos.inc"

extrn   printf  ; vararg functions are cdecl

%macro  dbg     1-*
        ; push args in reverse order, except the 1st
%rep    %0 - 1
%rotate -1
        push    %1
%endrep
        ; complete the arg rotation
%rotate -1
        ; push string on stack and jump past it
        call    %%next
        db      %1, 10, 0
        align   4
%%next:
        call    printf
        ; pop args
        add     esp, %0 * 4
%endmacro

;       ...
;       ...
;       ...

        dbg     "here%d-%d", 5, 7
        dbg     "here2"

**Comment by dborca** [comment]
Here is how I'd do the printf-debugging from inside ASM sources: http://pastie.org/10450076

```
%include "xos.inc"

extrn   printf  ; vararg functions are cdecl

%macro  dbg     1-*
        ; push args in reverse order, except the 1st
%rep    %0 - 1
%rotate -1
        push    %1
%endrep
        ; complete the arg rotation
%rotate -1
        ; push string on stack and jump past it
        call    %%next
        db      %1, 10, 0
        align   4
%%next:
        call    printf
        ; pop args
        add     esp, %0 * 4
%endmacro

;       ...
;       ...
;       ...

        dbg     "here%d-%d", 5, 7
        dbg     "here2"
```

**Comment by dborca** [comment]
... of course, you'd need pusha/popa. However, the printf itself may use the sse registers, so you might need to save sse/mmx registers as well...

**Comment by sezero** [comment]
There is a malloc_a facility in the source which sends back an address aligned at the requested boundary, however it is only called with 4 byte alignment requests.  IIRC (I may be wrong), SSE code would require to operate on addresses aligned at 16 bytes boundary?

**Comment by dborca** [comment]
Only a few select instructions need that alignment: movaps (aligned) vs movups (unaligned). But yes, you do have a point.

**Comment by dborca** [comment]
There is an alternative to malloc_a/free_a: posix_memalign/free. I do not know whether DJGPP (or windows, for that matter) has posix_memalign, though.

**Comment by sezero** [comment]
The SSE asm seems to be using both movaps and movups.

DJGPP has a memalign (and windows has _aligned_malloc). Never tried memalign of djgpp so don't know its usability.

Frank:  can you test changing the alinment param from 4 to 16 in the malloc_a() calls and retest to see if anything changes?

**Comment by dborca** [comment]
malloc_a takes a p2align, not a simple align. That is, malloc_a(size, 4) will align a block to 2^4=16.

But there might be places where I serendipitously got the right alignment on my then-configuration without enforcing it. That being said, I'd take another route: just replace every occurrence of movaps with movups and test again.

If it doesn't crash, there's your culprit. Afterwards, correct the code to use the right alignment and change back movups to movaps where possible (because they are faster).

**Comment by sezero** [comment]
> malloc_a takes a p2align, not a simple align. That is, malloc_a(size, 4) will align a block to 2^4=16.

Oh, you are right.  (Frank: joyously ignore my request.)

> just replace every occurrence of movaps with movups and test again.
> If it doesn't crash, there's your culprit. Afterwards, correct the code to
> use the right alignment and change back movups to movaps where possible
> (because they are faster).

Thanks.

**Comment by maraakate** [comment]
Will try replacing and report back.  Now there was a lot of signed vs 
unsigned warnings during compile time for x86 ASM as well from x86gen.c 
and drvgen.c.  Could this be affecting it?

**Comment by maraakate** [comment]
Changing movaps to movups works (so far).

Tried loading Quake 2 with and without packet driver (that was one of 
the hacky ways it worked).  Then tried Quake 1 (always requires param 
junk to start) and also tried quake 1 other params for loading mods 
(that would always break).

Seems to work OK.  Thanks so much.

Frank

**Comment by sezero** [comment]
Finally, some good news.  And how is performance affected by the change?

**Comment by maraakate** [comment]
There also appears to be no performance hit from exclusively using 
movups, but we can work towards making the code right.

Now, the other issue (as you saw from our spam last night... my 
apologies for this...) is the on-going windows issue with Quake 2 and 
wglChoosePixelFormat failing.  Now it appears you had it working at some 
point... was it tested on Windows 98SE?  I haven't setup a Windows XP 
machine install yet to see if that's part of the problem, but ghoul 
tried with a glide wrapper and had the same problems on a modern machine 
so I assume there is some sort of code rot?  We'll have to recompile 
that function with printfs to see why it's failing.

Frank

**Comment by dborca** [comment]
BTW, do you happen to know if DXE "linking" breaks the global var alignment constrains in any way?

I'd appreciate if you would put in a little extra effort and do The Right Thing (tm) aka fix this SSE problem properly :)

As for the other issue, well... it's an entirely different issue and I suggest filing it as a different bug. IIRC, I used sage mostly on 98SE to test/play. If I ever got it right, it was probably trial and error, as windows was never my OS of choice. Nor is it now, so I can't really help you there. I do have a few sage DLLs and source trees (about 10 of them) scattered on my backup HDD with small variations (probably experimental stuff) but the one I published seemed the most complete/stable.

**Comment by maraakate** [comment]
Well, we will work towards making the code proper as we are distributing 
this stuff so leaving in some hacky workarounds isn't something we try 
to do unless it's out of our hands to fix it (like the 2x init ctx's for 
sage for performance reasons, but we plan to look at that soon also).

By dxe linking you mean making ref_gl require gl.dxe with -P gl.dxe?  I 
only tested this with gl.dxe as that was where the problem was occuring 
with never being able to startup properly.

Good to know you were using 98SE, at least I'm testing on similar target 
stuff.  So you're saying the 0.9.2 you had on your defunct website is 
from the same source tree you posted publicly?

Also, I do recall reading from ancient posts AmigaMerlin gave you a V5 
6000.  Then I read some later thread on 3dfxzone.it that someone either 
bought or got one from you.  Was it the same card?  Was curious if you 
still had it and a machine that could still use it because I am very 
curious about hearing V5 6k performance on higher resolution stuff with 
these ports.  I saw there was some code for handling 4 chips added by 
you in glide3x and in DMesa so I'm assuming you were aiming for the code 
to at least work on that card.  From what I can understand, the 
performance only will change once you start bumping up the resolution 
and make the card work hard as there's a direct correlation for 
basically 1024x768 and lower in regards to nice performance gains and 
processor speed.  This is true of Windows IGL vs Mini drivers vs Sage vs 
Mesa vs all of it under DOS.

Frank

**Comment by maraakate** [comment]
But looks like I have more testing to do, because that init ctx 2x issue may be related to our memory corruption issue on SSE machines.  I never tested on the Pentium 1 to see if performance was degraded without the init ctx hack.

**Comment by sezero** [comment]
> BTW, do you happen to know if DXE "linking" breaks the global var alignment constrains in any way?

If you mean dxe3gen would mess var alignments when generating a dxe: I honestly don't know. Will have to check the dxe3gen source.

**Comment by sezero** [comment]
> > BTW, do you happen to know if DXE "linking" breaks the global var alignment constrains in any way?
> 
> If you mean dxe3gen would mess var alignments when generating a dxe: I honestly don't know. Will have to check the dxe3gen source.

You are more familiar with dxe3gen code as someone who actually wrote code for it.  Can you see anything fishy?
http://www.delorie.com/bin/cvsweb.cgi/djgpp/src/dxe/dxe3gen.c

**Comment by maraakate** [comment]
Okay, tested with -P gl.dxe omitted from the ref_gl and works as it should as well.  The Hunk_Alloc issue with static linked ref_gl + sage together still persists which is another issue altogether.  My guess there is that it could be a problem in ref_gl itself or Sage when used as a static library and certain things arent being freed properly when contexts are destroyed.  Will have to rebuild with debugging symbols and a minimal build to see if anything interesting there pops up.  Since static linked ref_gl + renderer choice is up to the user to compile themselves it's not as much of a problem, but may be aiding in finding a much deeper issue.

**Comment by dborca** [comment]
Yes, by dxe "linking" I meant the process of building the dxe image. I don't know what is ref_gl etc. Heck, I didn't even know that DXE3 is now the default instead of the old DXE1.

I only mentioned this because it's hard to believe I did not hit the SSE bug back in 2006, as just getting memory aligned by pure chance each and every time, is hard to believe. Besides memory allocations, there are only the global variables which could cause movaps faults.  It was just an idea.

LOL, no, I didn't write code _for_ DXE3 :) I took DXE2 from Andrew Zabolotny, changed it heavily (read: made it what it is today) and added it to DJGPP with some help from Charles W. Sandmann. Here: http://www.geocities.ws/dborca/djgpp/dxe3/dxe3.html
Yet, scouting for errors and/or design decisions in 12-year old hobby code -- even if most of them are mine -- is way beyond the scope of this issue.

Frank, I can't remember if the sage 0.9.2 DLL was built from this exact source tree. For some weird reasons, I can only find scattered source trees and DLLs on my old HDD, the CVS repo itself is nowhere to be found :(

Unfortunately, I cannot help you guys further; I do not have any desktops in my house nowadays. I do not have the time, either. And I can't remember the 4x and/or resolution code, it was probably mostly guess-work, as I had a shitty monitor back in the day.

That said, I'll try my best to help you fix outstanding bugs, like this SSE one. But I cannot do any serious work on these old projects of mine. Haven't you noticed my replies are peppered with "can't remember"? :) I really can't remember...

**Comment by maraakate** [comment]
I appreciate you taking the time to even review the issue at all, because you are no means obligated to on such an old, unsupported project.

With that said, ref_gl is the renderer DLL for Quake 2.  Since the game supports both a software and opengl renderer they made them DLLs so you can switch between them instead of having to distribute two separate binaries like they did previously with Quake 1.  Originally, we had ref_gl.dxe depend on gl.dxe for loading and sezero overhauled that so we load gl.dxe separately (to avoid an on-going issue with dlclose from child dxes and memory leaks).

The SSE bug and memory alignment issues is probably a DJGPP specific issue or even the choice of compiler version.  Not sure what existed in 2006, but we are using 3.46.

But in any case, do you remember your Voodoo 5 6000 tests?  Very curious to know how that card worked under DOS with SLI as you said you compiled a glquake build with mesa under DOS a very long time ago.

**Comment by sezero** [comment]
> Heck, I didn't even know that DXE3 is now the default instead of the old DXE1.

It is and q2dos very much relies on it

> Haven't you noticed my replies are peppered with "can't remember"? :) I really can't remember...

OK you made my day :)

**Comment by ghost** [comment]
To make a working win32 sage ICD, replace the contents of opengl32.def with the contents from pastebin: http://pastebin.com/xS1rSudf

Second, open the sage filename wgl.c and make the following change from the ORIGINAL:
/\* ICD functions **************************************************************/

#if 0

To this:
/\* ICD functions **************************************************************/

#if 1

**Comment by maraakate** [comment]
That's funny.  Well, if you can recompile and upload a new DLL to me on vogons I will test it out.  We sorted out the Hunk_Alloc and hardlock issue with chaos mod flashlight.  See q2dos bitbucket for more.  I think unless you have some information on the SSE issue we probably shouldn't spam this up too much more for the sake of Daniel's inbox :)  You can keep in touch with me on vogons or IRC or other methods.  Send me a message on vogons on how you want to stay in contact regarding work with Sage.

**Comment by ghost** [comment]
Wonderful, thank you!

**Comment by maraakate** [comment]
Got your message, but I can't reply back as your have private messages disabled.  I'd actually prefer email if you can do so... emoaddict15@gmail.com

**Comment by maraakate** [comment]
In any case, the game closes immediately as soon as I try to set opengl.  Even with SSE disabled with your win32 build.

**Comment by ghost** [comment]
I'll check the private msg settings.

Also confirmed that this quake2.exe binary works with the win32 sage ICD (with nglide):
http://www.vogons.org/download/file.php?id=19070

**Comment by maraakate** [comment]
To test your version I need the ref_gl.dll as well as the gamex86.dll.  Is this the original 3.20 version?

**Comment by maraakate** [comment]
@dborca You might be interested to know the performance of Sage (at least under DOS) is very close to AmigaMerlins OpenGL ICD.  I have a demo of an intense death match from a custom weapons mod that really puts some strain on the card and CPU.  At 640x480x16 I get ~57fps in Windows with AmigaMerlin and ~53fps in DOS with Sage.  Mesa has about the same numbers as well, ~53fps.  1024x768x16 has about the same speeds in this particular demo.

**Comment by sezero** [comment]
> To make a working win32 sage ICD, replace the contents of opengl32.def
> with the contents from pastebin: http://pastebin.com/xS1rSudf

That def is plain wrong: All the exported gl functions are GLAPIENTRY i.e. __stdcall  and that def removes the "@xx" __stdcall suffixes.

**Comment by dborca** [comment]
@sezero you can generate a def by hacking glapit.pl script file

@maraakate I think one could squeeze a little bit of speed from sage by forsaking the dream of a general-purpose GL implementation. That is, one could try to build GrVertex arrays directly instead of filling in TNL_VERTEXBUFFER arrays + copying them to the GrVertex.

I somehow planned to do that as a fork (but postponed) to keep sage open for a potential different hardware. I'm not sure how much faster it will get, or even if there are major roadblocks ahead, as it would require a bit of (or maybe major) of restructuring. NB: I'm talking from distant past memories, so take it with a grain of salt.

**Comment by ghoul8765** [comment]
![Uploading sage_ICD.PNG…]()

**Comment by maraakate** [comment]
@dborca I can't speak for @sezero but my coding abilities in regards to graphics programming are limited so that upgrade probably won't come from me... but you never know how I feel a few years from now.

**Comment by ghost** [comment]
There are a lot of statements with tnl_vb.

I wonder if Mesa3d has any commits which disable/enable TnL, even if just to benchmark the framerate difference.

Edit: http://sourceforge.net/p/mesa3d/mailman/message/5415943/

**Comment by maraakate** [comment]
@ghoul8765: That stuff is from 2000.  We're working with stuff from about 2004-2005.

**Comment by ghost** [comment]
Mainly benchmarking the cost of writing to TnL arrays instead of directly to video. According to Daniel's suggestion, this cost should be minimal, so perhaps it isn't worth pursuing.

These libraries, particularly mesa3d (or even sage) where coupled with nglide, could activate the direct3d pathway for the Windows Quake1/2 clients. This is useful where the video driver has an incomplete opengl implementation. The "quake series" forum at gog.com is actively discussing related issues.

[As an aside: been testing qdos, and perhaps an IPX=1 parameter at compile- or run-time would be useful where coupled with MAX_DATAGRAM=1024, as noted in your source code. Larger values are problematic because the corresponding buffer size is in conventional memory. Also noted that "reaper bot" mod does not display "scores" in console window where vanilla quake does. I could start by printing errors where the console window buffers overflow, if they don't already.]

**Comment by maraakate** [comment]
I don't know what you're talking about with nGlide but I have no plans to maintain this to work with glide wrappers.  The port is intended for real computers.  Emulators skip important things and can allow bad code to run without issues... This is OK when you're trying to play a game on an emulator, but on a real computer can blow up in your face.

I don't know about the reaper bot issues, you'll have to open up an issue on the qdos bitbucket issue tracker with more information.

**Comment by ghoul8765** [comment]
The win32 glquake build could use mesa3d/glide instead. The nglide wrapper would translate the glide to direct3d. This is one pathway to direct3d output on Windows clients without writing a direct3d driver for quake1/2, and enable computer systems to use the more compatible direct3d output.

**Comment by maraakate** [comment]
That would have to be a separate fork of the Q2DOS and friends projects you would have to maintain yourself as I am not interested in supporting those platforms.

**Comment by maraakate** [comment]
I'm also very confused on your proposal here.  Pretty much any card made after 2000-2001 can play Q1 and Q2 without too much trouble.  You might need to disable multitexturing, 8-bit palette textures and limit the gl_string length but otherwise they should play OK.

If you're that interested in a direct3d port look at d3dquake or some other port and merge that with QDOS together if you really require it's feature set.

And sezero and his buddies have spent a great deal of time refining quakespasm which is recommended if you want the expanded limits but a stock look.

Yamagi Quake 2 is recommended for the same reason.

**Comment by ghost** [comment]
I was thinking that the opengl quake versions sometimes aren't compatible with Windows video cards, such as some onboard video models. This is from recollection of threads I read at gog.com and elsewhere. Having the mesa/glide combination should allow a new user to run a Windows client, such as quakespasm, without working opengl on the video card. It would run direct3d through mesa/nglide. It would apply mainly for difficult (and rare?) cases where their opengl is broken and they wanted a workaround.

**Comment by maraakate** [comment]
You're talking about jumping through 3 layers to get a visual.  And most of those problems are intel cards and the specific problem is multitexturing and/or gl_dynamic 1.  Those cards usually can't handle it.

**Comment by ghost** [comment]
I didn't know that. It seems that they are trying random fixes at the forums. :)

Edit: switching from gcc484/djgpp205 to gcc453/djgpp204 fixed the reaper bot issue. Your default qdos binary didn't display the issue either (while testing mod by starting as single player since IPX=0).

**Comment by maraakate** [comment]
Did not see this edit, but we are using gcc346 and djgpp205.

**Comment by ghost** [comment]
I don't have the hardware to reproduce the performance issue in q2dos/sage which requires initializing the driver twice, but this may be worth testing:

``` diff
--- gl_sage.c   2016-03-19 20:38:46 -0400
+++ gl_sage-modified.c  2016-03-19 20:39:04 -0400
@@ -87,6 +87,7 @@ static int SAGE_InitCtx (int *width, int
    }

    if (sage_bind_fp(ctx, (void *)0xB16B00B5, *width, *height) != 0) {
+       sage_swap(1);
        sage_bind_fp (NULL, NULL, 0, 0);
        sage_shut_fp (ctx);
        sage_fini_fp ();
@@ -96,7 +97,7 @@ static int SAGE_InitCtx (int *width, int

    /* FS:  HACK HACK HACK:  For some reason SAGE has bad
     * performance unless it's loaded twice.  SLI or not. :( */
-   if (firstRunHack) {
+   if (0) {
        firstRunHack = false;
        ri.Con_Printf(PRINT_DEVELOPER, "Running SAGE performance hack\n");
        SAGE_Shutdown();
```

If not, then perhaps the initialization routine could be broken into parts and inserted in place of the "firstRunHack" section.

**Comment by maraakate** [comment]
It may be something that silly.  I can certainly test it out tomorrow and verify.

**Comment by maraakate** [comment]
Makes no difference; but thanks for the suggestion.

**Comment by ghost** [comment]
I appreciate this and your other tests of the library.

**Comment by sezero** [comment]
This discussion has gone way off-topic. As for the original issue, the source tree has a workaround for it. Closing.

**Comment by sezero** [comment]
For the record, this actually turned out to be a really stupid bug in q2dos engine,
and sage is not at fault.
